### PR TITLE
Add sync checker

### PR DESCRIPTION
**Pull request by gpeng**
Our existing migration sync checks don't support verifying correct unpublishing behaviour. Unfortunately, due to the way unpublishing works we need to check both content stores and have the document rather than the edition(s) available for the checks, so... back to the drawing board.

This is a new sync check framework that will allow us to compose new checks from some basic 'building blocks' (e.g. `TopLevelCheck`, `DetailsCheck`). It also abstracts away the unpublishing behaviour so as we don't need to consider it again for every `Edition` based format that we migrate.

Non-edition based formats can be supported too once we extract a common interface for a 'check' from `DetailedGuideCheck` 

The checks are not quite complete as there are a few elements of the `DetailedGuide` format that we either haven't covered yet (`parent`, `topics`, `related_mainstream_guides`) or else aren't quite working (`first_public_at`) due to other factors.

This is useful as is though and we will add support for the above in subsequent, hopefully more easily reviewable, PRs.

### CODE REVIEW

File: lib/sync_checker/checks/translations_check.rb [hunk 1]
```diff
@@ -0,0 +1,32 @@
+module SyncChecker
+  module Checks
+    TranslationsCheck = Struct.new(:expected_locales) do
+      attr_reader :content_item
+      def call(response)
+        stringified_locales = expected_locales.map(&:to_s)
```
Comment (boffbowsh) [inline lib/sync_checker/checks/translations_check.rb]: When you're calling this from `EditionBase`, it's possible for `expected_locales` to be `nil`. In that case, can we return early here?

File: test/unit/sync_checker/translations_check_test.rb [hunk 2]
```diff
@@ -0,0 +1,129 @@
+require 'minitest/autorun'
+require 'mocha/setup'
+require 'active_support'
+require 'active_support/json'
+require 'active_support/core_ext'
+require_relative '../../../lib/sync_checker/checks/translations_check'
+
+class TranslationsCheckTest < Minitest::Test
```
Comment (boffbowsh) [inline test/unit/sync_checker/translations_check_test.rb]: Ditto the above comment it'd be good to add a test to document what happens when `expected_locales` is nil

File: lib/sync_checker/formats/detailed_guide_check.rb [hunk 3]
```diff
@@ -6,6 +6,7 @@ def root_path
       end
 
       def checks_for_live(locale)
+        #related_mainstream
```
Comment (boffbowsh) [inline lib/sync_checker/formats/detailed_guide_check.rb]: You've committed some comments here

File: lib/sync_checker/formats/edition_base.rb [hunk 4]
```diff
@@ -128,17 +124,26 @@ def check_live(response, locale)
...
 
       def top_level_fields_hash(translation_for_locale)
         {
-          analytics_identifier: nil,
           base_path: get_path(translation_for_locale.locale),
           content_id: document.content_id,
           document_type: document_type,
           locale: translation_for_locale.locale.to_s,
           publishing_app: "whitehall",
-          rendering_app: rendering_app,
```
Comment (boffbowsh) [inline lib/sync_checker/formats/edition_base.rb]: Why aren't we testing this?

File: lib/sync_checker/checks/translations_check.rb (hunk 1 above)
Comment (gpeng) [inline lib/sync_checker/checks/translations_check.rb]: Do we care if the `expected_locales` passed to the constructor is `nil` but there are `available_translations` in the response?

File: lib/sync_checker/formats/detailed_guide_check.rb (hunk 3 above)
Comment (gpeng) [inline lib/sync_checker/formats/detailed_guide_check.rb]: fixed in 5c19a2e

File: lib/sync_checker/formats/edition_base.rb (hunk 4 above)
Comment (gpeng) [inline lib/sync_checker/formats/edition_base.rb]: Fixed in 14bae9f

**Comment by boffbowsh** [comment]
👍 for the bits I didn't do

File: lib/sync_checker/checks/translations_check.rb (hunk 1 above)
Comment (gpeng) [inline lib/sync_checker/checks/translations_check.rb]: I've removed the `.try` bit in `EditionBase` (in 99ef4be) as it is never nil. It shouldn't ever be `empty?` either. I think this is ok now and we can revise along they way if that proves not to be the case.

**Comment by boffbowsh** [comment]
@steventux short review please :)

File: test/unit/sync_checker/checks/details_check_test.rb [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+        response_code: 404,
+        body: {
+        }.to_json
+      )
+
+      expected = {
+        body: "a bit of html"
+      }
+
+      assert_equal [], DetailsCheck.new(expected).call(response)
```
Comment (steventux) [inline test/unit/sync_checker/checks/details_check_test.rb]: Do 404s and other unexpected responses get reported elsewhere? Is this considered different to not in sync or is it just noise you want to omit?

Comment (gpeng) [inline test/unit/sync_checker/checks/details_check_test.rb]: That's a good point. We've tried to make it so as an individual 'check' only returns errors if the thing it is checking is actually 'wrong' and returns none otherwise. We did have a specific `status` check at one point. We probably need to reinstate that.

Comment (gpeng) [inline test/unit/sync_checker/checks/details_check_test.rb]: I'm going to merge this so as we can get some feedback by running it. We have a couple of other checks to add before using it to make any migration 'decisions' in any case so I'll add the status check to the list.
